### PR TITLE
[v2] S3 high level checksums

### PR DESCRIPTION
**Pull request by aemous**
*Issue #, if available:*
- https://github.com/aws/aws-cli/issues/6750

Low-level `aws s3api` commands support checksums other than MD5 for verifying end-to-end data integrity. This pull request ports over similar functionality to high-level `aws s3` commands. 

*Description of changes:*
- Added `--checksum-algorithm` flag for file uploads using `aws s3 cp`, `aws s3 sync`, and `aws s3 mv`.
  - Supported algorithms are CRC32, SHA256, SHA1, CRC32C.
- Added `--checksum-mode` flag for file downloads using `aws s3 cp`, `aws s3 sync`, and `aws s3 mv`.
- Added unit and functional tests to assert the correctness of these changes.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: awscli/customizations/s3/utils.py [hunk 1]
```diff
@@ -519,6 +521,7 @@ def map_upload_part_params(cls, request_params, cli_params):
         """Map CLI params to UploadPart request params"""
         cls._set_sse_c_request_params(request_params, cli_params)
         cls._set_request_payer_param(request_params, cli_params)
+        cls._set_checksum_algorithm_param(request_params, cli_params)
```
Comment (hssyoo) [inline awscli/customizations/s3/utils.py:524]: Something interesting I noticed is that the `map_upload_part_params` isn't actually called anywhere. Based on the [commit history](https://github.com/aws/aws-cli/commit/fc25229aeebf3f13721b307fd658a57d3c617246#diff-a06cd12835dfbf1c3436863e220d13fc24a1fab17ace4de80cce8fa4b9158409), I think this is actually dead code that wasn't cleaned up after integration with `s3transfer`.

Digging into `s3transfer`'s upload logic, it looks like `UploadSubmissionTask` handles both regular and [multipart uploads](https://github.com/aws/aws-cli/blob/9419d473fdb7f9017ff1c935da5b5cd8834ce86f/awscli/s3transfer/upload.py#L696). And the AWS CLI's `UploadRequestSubmitter` [always sets `map_put_object_params` as the request mapper](https://github.com/aws/aws-cli/blob/v2/awscli/customizations/s3/s3handler.py#L341).

You can verify this by commenting out this line and calling `aws s3 cp large-file s3://my-bucket --checksum-algorithm CRC32` and observing that `CreateMultipartUpload` and `UploadPart` are called with the checksum header.

For this PR, let's remove this line but keep the method for now. I want to avoid polluting this PR with changes not totally relevant to the task. We can create a backlog item to address potential dead code.

Comment (hssyoo) [inline awscli/customizations/s3/utils.py:524]: Something interesting I noticed is that the `map_upload_part_params` isn't actually called anywhere. Based on the [commit history](https://github.com/aws/aws-cli/commit/fc25229aeebf3f13721b307fd658a57d3c617246#diff-a06cd12835dfbf1c3436863e220d13fc24a1fab17ace4de80cce8fa4b9158409), I think this is actually dead code that wasn't cleaned up after integration with `s3transfer`.

Digging into `s3transfer`'s upload logic, it looks like `UploadSubmissionTask` handles both regular and [multipart uploads](https://github.com/aws/aws-cli/blob/9419d473fdb7f9017ff1c935da5b5cd8834ce86f/awscli/s3transfer/upload.py#L696). And the AWS CLI's `UploadRequestSubmitter` [always sets `map_put_object_params` as the request mapper](https://github.com/aws/aws-cli/blob/v2/awscli/customizations/s3/s3handler.py#L341).

You can verify this by commenting out this line and calling `aws s3 cp large-file s3://my-bucket --checksum-algorithm CRC32` and observing that `CreateMultipartUpload` and `UploadPart` are called with the checksum header.

For this PR, let's remove this line but keep the method for now. I want to avoid polluting this PR with changes not totally relevant to the task. We can create a backlog item to address potential dead code.

File: awscli/customizations/s3/subcommands.py [hunk 2]
```diff
@@ -481,6 +481,16 @@
     )
 }
 
+CHECKSUM_MODE = {
+        'name': 'checksum-mode', 'choices': ['ENABLED'],
+        'help_text': 'To retrieve the checksum, this mode must be enabled.'
+}
+
+CHECKSUM_ALGORITHM = {
+        'name': 'checksum-algorithm', 'choices': ['CRC32', 'SHA256', 'SHA1', 'CRC32C'],
```
Comment (hssyoo) [inline awscli/customizations/s3/subcommands.py:490]: non-blocking: We should consider retrieving the choices from the service model instead of hardcoding the values here.

File: awscli/customizations/s3/subcommands.py [hunk 3]
```diff
@@ -1359,6 +1382,19 @@ def _raise_if_mv_same_paths(self, src, dest):
...
+        if paths_type != expected_paths_type:
+            expected_usage_map = {
+                'locals3': '<LocalPath> <S3Uri>',
+                's3s3': '<S3Uri> <S3Uri>',
+                's3local': '<S3Uri> <LocalPath>',
+                's3': '<S3Uri>'
+            }
+            raise ParamValidationError(
+                "The %s flag is only compatible with the format: %s. You used the format: %s." %
+                (param, expected_usage_map[expected_paths_type], expected_usage_map[paths_type])
```
Comment (hssyoo) [inline awscli/customizations/s3/subcommands.py:1395]:
```suggestion
                f"Expected {param} parameter to be used with one of following path formats: "
                f"{', '.join(allowed_paths)} but received {expected_usage_map[paths_type]}"
```
Where `allowed_paths` is a list of the mapped paths types.

File: awscli/customizations/s3/subcommands.py [hunk 4]
```diff
@@ -1359,6 +1382,19 @@ def _raise_if_mv_same_paths(self, src, dest):
                 f"{self.parameters['src']} - {self.parameters['dest']}"
             )
 
+    def _raise_if_paths_type_incorrect_for_param(self, param, paths_type, expected_paths_type):
+        if paths_type != expected_paths_type:
```
Comment (hssyoo) [inline awscli/customizations/s3/subcommands.py:1386]: Instead of `expected_paths_type` let's make the arg a list and call it  `allowed_paths_types`. And here we can make the check `if paths_type not in allowed_paths_types`.

File: awscli/customizations/s3/subcommands.py [hunk 5]
```diff
@@ -1276,6 +1286,19 @@ def _validate_path_args(self):
             if self._should_emit_validate_s3_paths_warning():
                 self._emit_validate_s3_paths_warning()
 
+        # If we're using the checksum-algorithm flag, the path type must be locals3
```
Comment (hssyoo) [inline awscli/customizations/s3/subcommands.py:1289]: We can remove this comment. It's just describing what the code is doing and doesn't have any explanatory benefits.

File: awscli/customizations/s3/subcommands.py [hunk 6]
```diff
@@ -1276,6 +1286,19 @@ def _validate_path_args(self):
...
+        if params.get('checksum_algorithm'):
+            self._raise_if_paths_type_incorrect_for_param(
+                CHECKSUM_ALGORITHM['name'],
+                params['paths_type'],
+                'locals3')
+        # If we're using the checksum-mode flag, the path type must be s3local
```
Comment (hssyoo) [inline awscli/customizations/s3/subcommands.py:1295]: Same here, we can remove.

File: awscli/customizations/s3/subcommands.py [hunk 7]
```diff
@@ -1276,6 +1286,19 @@ def _validate_path_args(self):
             if self._should_emit_validate_s3_paths_warning():
                 self._emit_validate_s3_paths_warning()
 
+        # If we're using the checksum-algorithm flag, the path type must be locals3
+        if params.get('checksum_algorithm'):
+            self._raise_if_paths_type_incorrect_for_param(
+                CHECKSUM_ALGORITHM['name'],
+                params['paths_type'],
+                'locals3')
```
Comment (hssyoo) [inline awscli/customizations/s3/subcommands.py:1294]: I think we should also allow `s3s3`. See my top-level comment for more info.

File: tests/unit/customizations/s3/test_utils.py [hunk 8]
```diff
@@ -653,6 +653,33 @@ def test_upload_part_copy(self):
              'SSECustomerKey': 'my-sse-c-key'})
 
 
+class TestRequestParamsMapperFlexibleChecksumsUploads:
```
Comment (hssyoo) [inline tests/unit/customizations/s3/test_utils.py:656]:
```suggestion
class TestRequestParamsMapperChecksumAlgorithm:
```

File: tests/unit/customizations/s3/test_utils.py [hunk 9]
```diff
@@ -653,6 +653,33 @@ def test_upload_part_copy(self):
...
+        RequestParamsMapper.map_put_object_params(request_params, cli_params)
+        assert request_params == {'ChecksumAlgorithm': 'CRC32'}
+
+    def test_upload_part(self, cli_params):
+        request_params = {}
+        RequestParamsMapper.map_upload_part_params(request_params, cli_params)
+        assert request_params == {'ChecksumAlgorithm': 'CRC32'}
+
+
+class TestRequestParamsMapperFlexibleChecksumsDownloads:
```
Comment (hssyoo) [inline tests/unit/customizations/s3/test_utils.py:672]:
```suggestion
class TestRequestParamsMapperChecksumMode:
```

File: tests/functional/s3/test_cp_command.py [hunk 10]
```diff
@@ -760,6 +760,72 @@ def test_cp_with_error_and_warning_permissions(self):
         self.assertIn('upload failed', stderr)
         self.assertIn('warning: File has an invalid timestamp.', stderr)
 
+    def test_upload_with_flexible_checksum_crc32(self):
```
Comment (hssyoo) [inline tests/functional/s3/test_cp_command.py:763]: Let's avoid using the term "flexible checksum". Instead, refer to the parameter names `checksum_algorithm` and `checksum_mode`. This is mainly to clarify what exactly the test is for.

**Review by hssyoo — CHANGES_REQUESTED** [review]
Looks like a solid start.

My main feedback here is that `--checksum-algorithm` should be supported for copies (ie `s3s3`). I can definitely see strong use cases for wanting to copy the object checksums between different S3 paths (or even generating different checksums). In order to support this, we're going to need to port [this PR](https://github.com/boto/s3transfer/pull/242) for the vended `s3transfer` package. Let's make this port a separate PR to reduce noise in this one.

We should also prefer f-strings for string formatting in all new code. ie `f"var: {my_var}"` instead of `"var: %s" % myvar`.

File: awscli/customizations/s3/subcommands.py (hunk 2 above)
Comment (aemous) [reply awscli/customizations/s3/subcommands.py:490]: Thanks for the feedback; could be worth looking into in a future PR.

File: tests/functional/s3/test_sync_command.py [hunk 11]
```diff
@@ -353,6 +353,64 @@ def test_with_copy_props(self):
             ]
         )
 
+    def test_upload_with_checksum_algorithm_sha1(self):
```
Comment (hssyoo) [inline tests/functional/s3/test_sync_command.py:356]: Let's also add copy (`s3s3`) tests here.

File: tests/unit/customizations/s3/test_subcommands.py [hunk 12]
```diff
@@ -400,6 +400,48 @@ def test_validate_no_streaming_paths(self):
         cmd_params.add_paths(paths)
         self.assertFalse(cmd_params.parameters['is_stream'])
 
+    def test_validate_checksum_algorithm_download_error(self):
+        paths = ['s3://bucket/key', self.file_creator.rootdir]
+        parameters = {'checksum_algorithm': 'CRC32'}
+        cmd_params = CommandParameters('cp', parameters, '')
+        with self.assertRaises(ParamValidationError):
```
Comment (hssyoo) [inline tests/unit/customizations/s3/test_subcommands.py:407]: `ParamValidationError` is used enough throughout the codebase that when testing we should assert that the error message is what we expect (and not propagating some other `ParamValidationError`).

File: tests/unit/customizations/s3/test_utils.py [hunk 13]
```diff
@@ -653,6 +653,28 @@ def test_upload_part_copy(self):
              'SSECustomerKey': 'my-sse-c-key'})
 
 
+class TestRequestParamsMapperChecksumAlgorithm:
```
Comment (hssyoo) [inline tests/unit/customizations/s3/test_utils.py:656]: Let's also add test cases for:
* Copy object
* Empty `cli_params` and assert `ChecksumAlgorithm` isn't present

File: tests/unit/customizations/s3/test_utils.py [hunk 14]
```diff
@@ -653,6 +653,28 @@ def test_upload_part_copy(self):
...
+    def cli_params(self):
+        return {'checksum_algorithm': 'CRC32'}
+
+    def test_put_object(self, cli_params):
+        request_params = {}
+        RequestParamsMapper.map_put_object_params(request_params, cli_params)
+        assert request_params == {'ChecksumAlgorithm': 'CRC32'}
+
+
+class TestRequestParamsMapperChecksumMode:
```
Comment (hssyoo) [inline tests/unit/customizations/s3/test_utils.py:667]: Let's add a test case for empty `cli_params` and assert `ChecksumAlgorithm` isn't present.

**Review by hssyoo — CHANGES_REQUESTED** [review]

File: tests/functional/s3/test_cp_command.py [hunk 15]
```diff
@@ -760,6 +760,66 @@ def test_cp_with_error_and_warning_permissions(self):
         self.assertIn('upload failed', stderr)
         self.assertIn('warning: File has an invalid timestamp.', stderr)
 
+    def test_upload_with_checksum_algorithm_crc32(self):
```
[the remark this answers is not in the document]
Comment (hssyoo) [reply tests/functional/s3/test_cp_command.py:763]: Oh, we should also be sure to add test cases for multipart uploads.

**Review by hssyoo — COMMENTED** [review]
Playing around with it some more, I'm not sure `--checksum-mode ENABLED` provides anything useful without the `debug` flag:
```
aws s3 cp s3://hssyoo-brawn-test/README.rst ./ --checksum-mode ENABLED

download: s3://hssyoo-brawn-test/README.rst to ./README.rst
```

Compare this to the low-level `get-object`:
```
aws s3api get-object --bucket hssyoo-brawn-test --key README.rst --checksum-mode ENABLED ./README.rst

{
    ...
    "ChecksumCRC32C": "ac1Isw==",
    ...
}
```

With the current changes, the `--checksum-mode` parameter in high-level S3 commands:
* Doesn't do client-side validation of the object
* Doesn't output the checksum the user could use to validate the object

We should decide what behavior we want to support.

EDIT: Sorry, validation is automatically [supported with `botocore`](https://github.com/aws/aws-cli/blob/5f942a86a529e48aa15fa959a55a24ab605cfa19/awscli/botocore/httpchecksum.py#L372). Ignore.

File: tests/functional/s3/test_cp_command.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (aemous) [reply tests/functional/s3/test_cp_command.py:763]: Resolving in next revision

File: tests/functional/s3/test_sync_command.py [hunk 16]
```diff
@@ -353,6 +353,64 @@ def test_with_copy_props(self):
...
+                'ContentLength': '100',
+                'LastModified': '00:00:00Z',
+                'ETag': 'foo-1',
+                'ChecksumSHA1': 'checksum',
+                'Body': BytesIO(b'foo')
+            }
+        ]
+        cmdline = f'{self.prefix} s3://bucket/foo {self.files.rootdir} --checksum-mode ENABLED'
+        self.run_cmd(cmdline, expected_rc=0)
+        self.assertEqual(self.operations_called[0][0].name, 'ListObjectsV2')
```
[the remark this answers is not in the document]
Comment (aemous) [reply tests/functional/s3/test_sync_command.py:397]: I ran the command manually with debug and verified that a download sync does not call GetObject. I also added a print statement to output all operations called when running this test and ListObjectV2 was the only one.

File: tests/functional/s3/test_sync_command.py (hunk 11 above)
Comment (aemous) [reply tests/functional/s3/test_sync_command.py:356]: Good idea thanks, adding one in next revision

File: tests/functional/s3/test_sync_command.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (hssyoo) [reply tests/functional/s3/test_sync_command.py:397]: But not having `GetObject` defeats the purpose of this test since we're explicitly testing the `ChecksumMode` parameter with `sync`'s download feature. It looks like the issue here is that we have `self.head_object_response()` as the stubbed response for `ListObjectsV2` when it should be `self.list_objects_response(keys['foo-1'])` instead.

File: tests/unit/customizations/s3/test_subcommands.py (hunk 12 above)
Comment (aemous) [reply tests/unit/customizations/s3/test_subcommands.py:407]: Good point, resolving in next revision

File: tests/unit/customizations/s3/test_utils.py (hunk 13 above)
Comment (aemous) [reply tests/unit/customizations/s3/test_utils.py:656]: Good points, resolving in next iteration

File: tests/unit/customizations/s3/test_utils.py (hunk 14 above)
Comment (aemous) [reply tests/unit/customizations/s3/test_utils.py:667]: Good point thank you, resolving in next revision

File: tests/unit/customizations/s3/test_subcommands.py [hunk 17]
```diff
@@ -400,6 +400,42 @@ def test_validate_no_streaming_paths(self):
         cmd_params.add_paths(paths)
         self.assertFalse(cmd_params.parameters['is_stream'])
 
+    def test_validate_checksum_algorithm_download_error(self):
+        paths = ['s3://bucket/key', self.file_creator.rootdir]
+        parameters = {'checksum_algorithm': 'CRC32'}
+        cmd_params = CommandParameters('cp', parameters, '')
+        with self.assertRaises(ParamValidationError) as cm:
+            cmd_params.add_paths(paths)
+            self.assertIn('Expected checksum-algorithm parameter to be used with one of following path formats', cm.msg)
```
Comment (hssyoo) [inline tests/unit/customizations/s3/test_subcommands.py:409]: Let's also make this assertion for the tests below.

**Review by hssyoo — CHANGES_REQUESTED** [review]

File: awscli/customizations/s3/subcommands.py [hunk 18]
```diff
@@ -1359,6 +1380,19 @@ def _raise_if_mv_same_paths(self, src, dest):
...
+        if paths_type not in allowed_paths:
+            expected_usage_map = {
+                'locals3': '<LocalPath> <S3Uri>',
+                's3s3': '<S3Uri> <S3Uri>',
+                's3local': '<S3Uri> <LocalPath>',
+                's3': '<S3Uri>'
+            }
+            raise ParamValidationError(
+                f"Expected {param} parameter to be used with one of following path formats: "
+                f"{', '.join(allowed_paths)} but received {expected_usage_map[paths_type]}"
```
[the remark this answers is not in the document]
Comment (aemous) [reply awscli/customizations/s3/subcommands.py:1393]: Ah, will fix in next revision, with a tiny adjustment to grammar/wording to be more clear.

File: .changes/next-release/feature-s3-46667.json [hunk 19]
```diff
@@ -0,0 +1,5 @@
+{
+  "type": "feature",
+  "category": "s3",
+  "description": "Adds support for checksums other than MD5 to high-level S3 commands."
```
[the remark this answers is not in the document]
Comment (aemous) [reply .changes/next-release/feature-s3-46667.json:4]: Will include in next revision

File: tests/unit/customizations/s3/test_subcommands.py (hunk 17 above)
Comment (aemous) [reply tests/unit/customizations/s3/test_subcommands.py:409]: Will do in next revision

**Review by hssyoo — APPROVED** [review]
Looks good! 🏆
